### PR TITLE
Partial clean up of copy & paste code for xthin stats

### PR DESCRIPTION
**Pull request by awemany**
There's a lot of repeated code to deal with expiry of old
statistics values after a day.

This commit replaces the repeated sections with a call to a
template function to do so.

### CODE REVIEW

**Comment by awemany** [comment]
Note: This is a WIP and for gathering feedback, as at least a good test case is missing.

Does anyone have a good idea on how to implement a test without complicating the code too much? The main issue is the time-dependency here.

**Comment by awemany** [comment]
Added test case and did further clean-up.

**Comment by awemany** [comment]
Added test case and further refactoring of average calculation.

**Comment by sickpig** [comment]
@awemany 
the failure on build 1670.1 (https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/builds/233185173) 
seems unrelated to your last commit 7056ddc. the failure is on`wallet.py`: 
https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/233218158#L2136

restarting the job since it seems a spurios failure.

**Comment by awemany** [comment]
I think this is ready for review now.

**Comment by awemany** [comment]
Rebased to just a single proper commit.

File: src/thinblock.cpp [hunk 1]
```diff
@@ -1341,15 +1244,15 @@ bool CThinBlockData::CheckThinblockTimer(uint256 hash)
...
         LogPrint("thin", "Starting Preferential Thinblock timer\n");
     }
     else
     {
         // Check that we have not exceeded the 10 second limit.
         // If we have then we want to return false so that we can
         // proceed to download a regular block instead.
-        uint64_t elapsed = GetTimeMillis() - mapThinBlockTimer[hash];
+        uint64_t elapsed = getTime() - mapThinBlockTimer[hash];
         if (elapsed > 10000)
```
Comment (ptschip) [inline src/thinblock.cpp]: I don't think this won't work, the timer will never be exceeded since it is set at 10000 millis.  It seems also a bit out of scope for this PR, since this PR is about Thinstats and not the thinblocktimer?

Comment (awemany) [reply src/thinblock.cpp]: Oops. Fuck. Yes, that was not meant to be that way. You are absolutely right. I did a search & replace and then noted that that went too far (into non-stats code) and took that back and apparently this one slipped through. Very good catch, will fix.

Comment (awemany) [reply src/thinblock.cpp]: I also noted that I named getTime() badly as there is GetTime() as well. Will rename that one now to getTimeForStats() as well.

Comment (awemany) [reply src/thinblock.cpp]: Also removed  LOCK(cs_thinblockstats); from ::updateStats() as that was superfluous.

**Comment by awemany** [comment]
Ok, also added AssertLockHeld() to newly defined methods, as per @ptschip's comment on slack.

**Comment by ptschip** [comment]
@awemany All looks good regarding code..I'll run this for 24 hrs (without the Clear() in the stats constructor) just to be sure before approving.

**Review by ptschip — APPROVED** [review]
tested and all looks good.
